File: NODES/Split System/Wrappers/Java/src/com/forgeNode/base/NODE.java
/*
*************** Name: Forge Node
*************** Part of: Java Wrapper libraries for nodes of split system
*************** Author: Aleksandar Damnjanovic AKA Kind Spirit
*************** YouTube channel: Kind Spirit Technology
*************** Date: 20.8.2026.
*************** Location: Kragujevac, Serbia
*/

package com.forgeNode.base;

import java.util.ArrayList;

public class NODE {
    private int nodeIndex;
    private int numOfSwitches;
    private int numOfSensors;
    private int numOfTransmitters;

    private String serverIP;
    private int port;

    private boolean clientNodeSwitches[];
    private float clientNodeSensors[];
    private String clientNodeTransmitters[];

    /* 
        Node can be accessed directly only if outside of the process.
        When process is running, node can be accessed through the process, otherwise
        you are risking synchronization issue
    */
    public NODE(int nodeIndex, int numOfSwitches, int numOfSensors, int numOfTransmitters,
        String serverIP, int port){

        this.nodeIndex = nodeIndex;
        this.numOfSwitches = numOfSwitches;
        this.numOfSensors = numOfSensors;
        this.numOfTransmitters = numOfTransmitters;

        this.serverIP = serverIP;
        this.port = port;

        boolean switches[]= new boolean[this.numOfSwitches];
        for(int i = 0 ; i < this.getNumOfSwitches(); i++)
            switches[i] = false;

        float sensors[]= new float[this.numOfSensors];
        for(int i = 0 ; i < this.getNumOfSensors(); i++)
            sensors[i] = 0;

        String transmitters[]= new String[this.numOfTransmitters];
        for(int i = 0 ; i < this.getNumOfTransmitters(); i++)
            transmitters[i] = "";

        this.clientNodeSwitches = switches;
        this.clientNodeSensors = sensors;
        this.clientNodeTransmitters= transmitters;
    }

    public NODE(int nodeIndex, int numOfSwitches, int numOfSensors, int numOfTransmitters,
        String serverIP, int port, boolean clientNodeSwitches[],
        float clientNodeSensors[], String clientNodeTransmitters[]){

        this.nodeIndex = nodeIndex;
        this.numOfSwitches = numOfSwitches;
        this.numOfSensors = numOfSensors;
        this.numOfTransmitters = numOfTransmitters;

        this.serverIP = serverIP;
        this.port = port;
        this.clientNodeSwitches = clientNodeSwitches;
        this.clientNodeSensors = clientNodeSensors;
        this.clientNodeTransmitters= clientNodeTransmitters;
    }

    public String getMessage(){
        String message= new String();
        for(int i = 0; i < getNumOfSwitches(); i++)
            message = message + "Cr_" + getNodeIndex() + "_C_" + i + "_" + (getSwitchValue(i)?"1":"0") + " ";
        
        for(int i = 0; i < getNumOfSensors(); i++)
            message = message + "Cr_" + getNodeIndex() + "_R_" + i + "_" + String.format("%.2f", getSensorValue(i)) + " ";
        
        message = message + "\n";

        return message;
    }

    public String generateInstructions(){
        String message= new String();
        for(int i = 0; i < getNumOfSwitches(); i++)
            message = message + "I_" + getNodeIndex() + "_C_" + i + "_" + (getSwitchValue(i)?"1":"0") + " ";
        
        for(int i = 0; i < getNumOfTransmitters(); i++)
            message = message + "I_" + getNodeIndex() + "_T_" + i + "_" + getTransmitterValue(i)+ " ";
        
        message = message + "\n";

        return message;
    }

    public void processReport(String rawMessage){
        ArrayList<NODE_INSTRUCTION> lista= NODE_INSTRUCTION.getListOfInstructions(rawMessage);
        for(int i = 0; i < lista.size(); i++){
            if(lista.get(i).getNodeIndex() != this.getNodeIndex())
                continue;
            else{
                if(lista.get(i).getType()=='R')
                    this.writeSensor(lista.get(i).getElementIndex(), lista.get(i).getFloatValue());
            }
        }
    }

    public void processInstructions(String rawMessage){
        ArrayList<NODE_INSTRUCTION> lista= NODE_INSTRUCTION.getListOfInstructions(rawMessage);
        for(int i = 0; i < lista.size(); i++){
            if(lista.get(i).getNodeIndex() != this.getNodeIndex())
                continue;
            else{
                if(lista.get(i).getType()=='C')
                    this.writeSwitch(lista.get(i).getElementIndex(), (lista.get(i).getIntValue()==1? true : false));
                else if(lista.get(i).getType()=='T')
                    this.writeTransmitter(lista.get(i).getElementIndex(), lista.get(i).getStringValue());
            }
        }
    }

    public int getNodeIndex() {
        return nodeIndex;
    }

    public int getNumOfSwitches() {
        return numOfSwitches;
    }

    public int getNumOfSensors() {
        return numOfSensors;
    }

    public int getNumOfTransmitters() {
        return numOfTransmitters;
    }

    public String getServerIP() {
        return serverIP;
    }

    public int getPort() {
        return port;
    }

    public boolean[] getClientNodeSwitches() {
        return clientNodeSwitches;
    }

    public float[] getClientNodeSensors() {
        return clientNodeSensors;
    }

    public String[] getClientNodeTransmitters() {
        return clientNodeTransmitters;
    }

    public void writeSwitch(int index, boolean value){
        if(index < this.getNumOfSwitches()  && index > -1)
            this.getClientNodeSwitches()[index]= value;
    }

    public void writeSensor(int index, float value){
        if(index < this.getNumOfSensors()  && index > -1)
            this.getClientNodeSensors()[index]= value;
    }

    public void writeTransmitter(int index, String value){
        if(index < this.getNumOfTransmitters()  && index > -1)
            this.getClientNodeTransmitters()[index]= value;
    }

    public boolean getSwitchValue(int index){
        if(index < this.getNumOfSwitches() && index > -1)
            return this.getClientNodeSwitches()[index];
        else
            return false;
    }

    public float getSensorValue(int index){
        if(index < this.getNumOfSensors() && index > -1)
            return this.getClientNodeSensors()[index];
        else
            return -1;
    }

    public String getTransmitterValue(int index){
        if(index < this.getNumOfTransmitters() && index > -1)
            return this.getClientNodeTransmitters()[index];
        else
            return "";
    }

}

File: NODES/Wrappers/python/NODE.py
'''
*************** Name: Forge Node
*************** Part of: Python Wrapper libraries
*************** Author: Aleksandar Damnjanovic AKA Kind Spirit
*************** YouTube channel: Kind Spirit Technology
*************** Date: 22.8.2026.
*************** Location: Kragujevac, Serbia
'''

class NODE:
    index = -1

    numOfSwitches = 0
    numOfSensors = 0
    numOfTransmitters = 0

    switches = None
    sensors = None
    transmitters = None

    def __init__(this, index, numOfSwitches, numOfSensors, numOfTransmitters, 
            switches: list[bool], sensors: list[float], transmitters: list[str]):
        this.index = index
        this.numOfSensors = numOfSensors
        this.numOfSwitches = numOfSwitches
        this.numOfTransmitters = numOfTransmitters
        this.switches = switches
        this.sensors = sensors
        this.transmitters = transmitters

    def generateMessage(this):
        message = ""
        for i in range(0, this.numOfSwitches):
            val = 0
            if this.switches[i]:
                val = 1
            message = f"{message}Cr_{this.index}_C_{str(i)}_{str(val)} "

        for i in range(0, this.numOfSensors):
            message = f"{message}Cr_{this.index}_R_{str(i)}_{str(this.sensors[i])} "

        return message

    def processInstructions(this, instructions: str):
        tokens = instructions.split(" ")

        for t in tokens:
            e = t.split("_")

            if e[0]!= 'I':
                print(f"Instruction {instructions}, not properly formated\n")
                return

            if e[1]!= str(this.index):
                print(f"Instruction {instructions}, holds wrong node index\n")
                return

            if e[2]=='C':
                val = bool(e[4])
                this.switches[int(e[3])] = val
            elif e[2]=='T':
                this.transmitters[int(e[3])] = str(e[4])

    def getSwitchValue(this, elIndex):
        return this.switches[elIndex]

    def getSensorValue(this, elIndex):
        return this.sensors[elIndex]

    def getTransmitterValue(this, elIndex):
        return this.transmitters[elIndex]

    def setSensorValue(this, elIndex, value: float):
        this.sensors[elIndex]= value